pkg/controller/clusterautoscaler/monitoring.go

```diff
@@ -0,0 +1,210 @@
+package clusterautoscaler
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1"
+	autoscalingv1 "github.com/openshift/cluster-autoscaler-operator/pkg/apis/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// ensureObjectForCA will ensure an object is created or updated according to the passed f mutate function
+func (r *Reconciler) ensureObjectForCA(ca *autoscalingv1.ClusterAutoscaler, desired metav1.Object, f controllerutil.MutateFn) (controllerutil.OperationResult, error) {
+	if err := controllerutil.SetControllerReference(ca, desired, r.scheme); err != nil {
+		return "", err
+	}
+
+	ro, ok := desired.(runtime.Object)
+	if !ok {
+		return "", fmt.Errorf("can not covert %T to a runtime.Object", desired)
+	}
+
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, ro, f)
+	if err != nil {
+		return "", err
+	}
+
+	klog.V(4).Infof("Ensuring object %q of type %T, operation: %v", desired.GetName(), desired, op)
+	return op, nil
+}
+
+// ensureAutoscalerServiceMonitor will create a service
+// for the given ClusterAutoscaler custom resource instance.
+func (r *Reconciler) ensureAutoscalerService(ca *autoscalingv1.ClusterAutoscaler) (controllerutil.OperationResult, error) {
```

Comment:
We use "ensure" but I think it would be clearer if it hinted at what it attempts to "ensure" - `ensureAutoscalerServiceMonitor{Exists,Created,SomethingElse}`? Ditto for the other following ensure-based functions. Edit: having looked at the related functions do we want to s/ensure/create, or s/ensure/createOrUpdate?